lib/srv/alpnproxy/local_proxy.go

```diff
@@ -303,3 +316,92 @@ func (l *LocalProxy) Close() error {
 	}
 	return nil
 }
+
+// StartAWSAccessProxy starts the local AWS CLI proxy.
+func (l *LocalProxy) StartAWSAccessProxy(ctx context.Context) error {
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{
+			NextProtos:         []string{string(l.cfg.Protocol)},
+			InsecureSkipVerify: l.cfg.InsecureSkipVerify,
+			ServerName:         l.cfg.SNI,
+			Certificates:       l.cfg.Certs,
+		},
+	}
+	proxy := &httputil.ReverseProxy{
+		Director: func(outReq *http.Request) {
+			outReq.URL.Scheme = "https"
+			outReq.URL.Host = l.cfg.RemoteProxyAddr
+		},
+		Transport: tr,
+	}
+	err := http.Serve(l.cfg.Listener, http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		if err := l.verifyAWSSignature(req); err != nil {
+			rw.WriteHeader(http.StatusForbidden)
+			return
+		}
+		proxy.ServeHTTP(rw, req)
+	}))
+	if err != nil && !utils.IsUseOfClosedNetworkError(err) {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// verifyAWSSignature verifies the request signature ensuring that the request originates from tsh aws command execution
+// AWS CLI signs the request with random generated credentials that are passed to LocalProxy by
+// the AWSCredentials LocalProxyConfig configuration.
+func (l *LocalProxy) verifyAWSSignature(req *http.Request) error {
+	sigV4, err := appaws.ParseSigV4(req.Header.Get("Authorization"))
+	if err != nil {
+		return trace.BadParameter(err.Error())
+	}
+	// Read the request body and replace the body ready with a new reader that will allow reading the body again
+	// by HTTP Transport.
+	payload, err := appaws.GetAndReplaceReqBody(req)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	reqCopy := req.Clone(context.Background())
+
+	// Remove all the headers that are not present in awsCred.SignedHeaders.
+	filterSignedHeaders(reqCopy, sigV4.SignedHeaders)
+
+	// Get the date that was used to create the signature of the original request
+	// originated from AWS CLI and reuse it as a timestamp during request signing call.
+	t, err := time.Parse(appaws.AmzDateTimeFormat, reqCopy.Header.Get(appaws.AmzDateHeader))
+	if err != nil {
+		return trace.BadParameter(err.Error())
+	}
+
+	signer := v4.NewSigner(l.cfg.AWSCredentials)
+	_, err = signer.Sign(reqCopy, bytes.NewReader(payload), sigV4.Service, sigV4.Region, t)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	localSigV4, err := appaws.ParseSigV4(reqCopy.Header.Get("Authorization"))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Compare the origin request AWS SigV4 signature with the signature calculated in LocalProxy based on
+	// AWSCredentials taken from LocalProxyConfig.
+	if sigV4.Signature != localSigV4.Signature {
+		return trace.AccessDenied("signature verification failed")
+	}
+	return nil
+}
+
+// filterSignedHeaders removes request headers that are not in the signedHeaders list.
+func filterSignedHeaders(r *http.Request, signedHeaders []string) {
```

Comment:
Nit: This is generic enough that I'd consider renaming it to just `filterHeaders` and placing somewhere in utils.